docs(api): Refactor instrument-context docstrings

jwwojak opened the pull request:
# Overview

This PR is for managing suggested revisions to the docstrings in `instrument_context.py`. This is descriptive, customer/user-facing text that appears in [Protocols and Instruments section](https://docs.opentrons.com/v2/new_protocol_api.html) of the API v2 reference. JIRA issue is RTC-204.

These are lightweight changes that revise text, grammar, style, punctuation. Changes also include anchors to longer, more expository API docs throughout because many of the docstrings are terse and/or brief. Linking to other docs helps provide readers with more context, information, code samples, and broader understanding of our API methods. In all cases, we've tried to change as little text as possible.

Excluded: any text related to warnings and errors (``raises``, ``runtime``, ``type``, etc.) or anything that is a combination of static text and dynamic code. Just revisions to main definitions text only.

# Test Plan

Sandbox: http://sandbox.docs.opentrons.com/instrument-context-docstrings/v2/new_protocol_api.html#

Testing consisted of making text revisions, pushing to the working branch, and reviewing text results in the sandbox.

# Changelog

Definition text changes throughout. Things like:

- Align text with corresponding method descriptions in the larger API docs.
- Adding anchor links to other sections.
- Punctuation (sentences ended without periods)
- Replacing units of measurement in text with abbreviations (µL replaces "microliter") and getting measurements styled properly (e.g. "20 mm" replaces "20mm"). 
- Really tried to be as conservative/light-touch as possible throughout.

# Review requests

A lot of this is small and intricate. Any second (or third or _n_) set of eyes reviewing helps.

# Risk assessment

Low, docs/text change only, at least I hope I've only changed text and not working code. heh.

jwwojak committed:
Spelling check

Line 116, "millimiters" should be "millimeters".

jwwojak committed:
fix microliters abbreviation-use mu not u

Use "µL" instead of "uL".

jwwojak committed:
starting_tips, reset_tipracks, default_speed

jwwojak committed:
reset_tipracks description

jwwojak committed:
Sandbox test

jwwojak committed:
sandbox test

ignore

jwwojak committed:
aspirate and assoc arguments

jwwojak committed:
Aspirate and This instance

Inconsistent punctuation for the text ":returns: This instance." Sometimes ended with a period, sometimes not. Changed to end all instances w/ a period.

jwwojak committed:
blow_out

jwwojak committed:
blow_out code formatting to monospace

jwwojak committed:
configure_for_volume and e.g.

Too many instances of "eg". Should be "e.g.," or "e.g."

jwwojak committed:
configure_for_volume part 2

jwwojak committed:
current_volume, channels, default_speed

jwwojak committed:
dispense description-definitions

jwwojak committed:
dispense format testing

Not rendering in a nice way. Might be exceeding character count per line.

jwwojak committed:
dispense formatting test

Again, more line length testing.

jwwojak committed:
dispense format layout test

jwwojak committed:
drop tip and new anchor

Anchor to pipette_tips.rst, "Dropping a Tip" section. Anchor is .. _pipette-drop-tip:

jwwojak committed:
drop_tip text formatting

jwwojak committed:
drop_tip formatting yet again

jwwojak committed:
Punctuation, links, vol measurement units

jwwojak committed:
Mix

jwwojak committed:
mix text formatting

jwwojak committed:
Mix: more formatting

jwwojak committed:
Mix one more format try

jwwojak committed:
Adding an anchor

jwwojak committed:
air_gap volume text

Last minute, saw some weird original text to fix.

jwwojak committed:
Check and fix any remaining

Returning to this after problems last week. Just a few additional changes.

Ran Black on the file. Returns the following:

  joewojak@admins-Air-2 protocol_api % black
  instrument_context.py --diff --color
  All done! ✨ 🍰 ✨
  1 file would be left unchanged.

jwwojak committed:
Merge branch 'edge' into instrument-context-docstrings

jwwojak commented:
@ecormany 

Picking up and trying again. Ran Black. Output suggests this might be ok:

> joewojak@admins-Air-2 protocol_api % black
  instrument_context.py --diff --color
  All done! ✨ 🍰 ✨
  1 file would be left unchanged.

Also, resolved small text conflict around `air_gap`.

jwwojak committed:
Line 618 remove spaces, keep same line

Line count doesn't change

jwwojak commented:
Changing status from draft.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
I think, in the spirit of "write the introduction last", it would be good to come back and revise the docstring for the class as a whole (lines 58–71) as well now.

jwwojak replied:
Took a run at a generic intro for this section. It might be kind of "meh." I figure most readers _aren't_ going to rely on it to find what they need. Instead, they may get to individual class objects via search or from links on other API docs.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -170,10 +171,9 @@ def aspirate(
-                     The flow rate for this aspirate will be
-                     `rate` * :py:attr:`flow_rate.aspirate <flow_rate>`.
-                     If not specified, defaults to 1.0.
+        :param rate: How quickly a pipette aspirates liquid, measured in µL/s.
```
This parameter is *not* measured in µL/s. 200 sets it to 200x the default rate, not 200 µL/s. 

The old copy is a little clunky though.
```suggestion
        :param rate: A multiplier for the default flow rate of the pipette.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -170,10 +171,9 @@ def aspirate(
-                     `rate` * :py:attr:`flow_rate.aspirate <flow_rate>`.
-                     If not specified, defaults to 1.0.
+        :param rate: How quickly a pipette aspirates liquid, measured in µL/s.
+                     Calculated as `rate` * :py:attr:`flow_rate.aspirate <flow_rate>`.
```
Picky, but I'd love to get rid of the unformatted asterisk indicating multiplication here. (It would be fine in a code snippet, but it's not). `times` or `multiplied by` would suffice.

jwwojak replied:
Fixing. Will use "multiplied by".  Found multiple instances and some hidden single quote for code instead of ``.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -256,35 +256,37 @@ def dispense(
+                        If ``location`` is a :py:class:`.Well`,
+                        the robot will dispense into
+                        :py:obj:`well_bottom_clearance.dispense <well_bottom_clearance>`
+                        mm above the bottom of the well.
```
"into clearance millimeters above the bottom of the well" has me 😵 …I don't think it's even grammatical. Maybe: 
```suggestion
                        If ``location`` is a :py:class:`.Well`,
                        the pipette will dispense at or above the center bottom of the well.
                        The distance from the bottom is specified by                 
                        :py:obj:`well_bottom_clearance.dispense <well_bottom_clearance>`.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -256,35 +256,37 @@ def dispense(
+                        :py:obj:`well_bottom_clearance.dispense <well_bottom_clearance>`
+                        mm above the bottom of the well.
+                        If ``location`` is  :py:class:`.Location`
+                        (i.e. the result of :py:meth:`.Well.top` or
```
```suggestion
                        (e.g., the result of :py:meth:`.Well.top` or
```
this list is non-exhaustive. and a comma. 🤓

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -256,35 +256,37 @@ def dispense(
+                        If ``location`` is  :py:class:`.Location`
+                        (i.e. the result of :py:meth:`.Well.top` or
+                        :py:meth:`.Well.bottom`), the robot will dispense into
+                        the specified position. If only a ``location`` is passed
```
I think the sentence that starts on this line needs to go last, otherwise the antecedent of the null pronoun in `If unspecified...` gets totally lost.

jwwojak replied:
In the words of the inimitable Mr. Torgue, "[That sentence has too many syllables](https://youtu.be/UJj85ToIC7Y?feature=shared&t=10). Apologize!"  😄

jwwojak replied:
Re-wrote `:param location:` for `aspirate` and `dispense`. Make descriptions more parallel. `aspirate` was a bit different. I think this is better than before.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -256,35 +256,37 @@ def dispense(
+                        will be dispensed (amount is accessible through
+                        :py:attr:`current_volume`). If unspecified,
+                        the robot will dispense into the current position.
+        :param rate: How quickly a pipette dispenses liquid. Measured in µL/s.
```
same as for `aspirate()`

jwwojak replied:
Rewrote to so `aspirate` and `dispense` are similar.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -256,35 +256,37 @@ def dispense(
         :type rate: float
-        :param push_out: Continue past the plunger bottom to guarantee all liquid
-                        leaves the tip. Specified in microliters. By default, this value is None.
+        :param push_out: Continue past the plunger bottom to help ensure all liquid
```
👍 under-promise, over-deliver.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
note at the end of `dispense()`: we could update it to match our style for including parentheses in method references, and to use `pipette` as our default `InstrumentContext` object in examples.

jwwojak replied:
Fixing `dispense()`.

Keeping as reminder to check `aspirate()`.

jwwojak replied:
Revising `aspirate` to match `dispense` in the Note callout.

**Dispense**
> If ``dispense`` is called with a single argument, it will not try
            to guess whether the argument is a volume or location - it is
            required to be a volume. If you want to call ``dispense`` with only
            a location, specify it as a keyword argument:
            ``pipette.dispense(location=plate['A1'])``

**Aspirate**
> If ``aspirate`` is called with a single argument, it will not try
            to guess whether the argument is a volume or location - it is
            required to be a volume. If you want to call ``aspirate`` with only
            a location, specify it as a keyword argument:
            ``pipette.aspirate(location=plate['A1'])``

For `dispense`, we also discussed shortening the location code example in the text for both methods:

Change:
> (as in instr.dispense(location=plate['A1']))

To:
> (e.g., (location=plate['A1']))

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -377,25 +379,26 @@ def mix(
+                         (e.g, ``plate.rows()[0][0].bottom()``).
+                         If unspecified, the pipette will mix from
+                         its current position.
         :type location: types.Location
```
this seems wrong. it's Well or Location. and it doesn't render anyway, so we should probably remove the line.

jwwojak replied:
Removing line 390.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -125,20 +124,21 @@ def starting_tip(self, location: Union[labware.Well, None]) -> None:
         default, the speed of individual motions can be changed with the
-        ``speed`` argument to :py:meth:`InstrumentContext.move_to`.
+        ``speed`` argument of the :py:meth:`InstrumentContext.move_to` method.
+        See also, :ref:`gantry_speed`.
```
super nit, i don't think there should ever be a comma after `See also`.

jwwojak replied:
A great topic for discussion, and a mushy one as this might be more style than rules.

We ought to settle on this as a style - comma or no comma. I've always used `See<comma> rest of sentence` as if it were a signal like "For example, . . . " so "See,  . . . " or "See also, . . . " is ok. To me it's the difference between mashing things together or separating them:

- SeeTitleofThingHere
vs
- See `<comma pause>` Title of Thing Here.

I think lawyerly types may not use this much e.g., "See _name of case law thing here_", but we're not doing legal writing.

This is one of those things like using a comma with i.e. or e.g.  Either way is acceptable, IMO.

- i.e.,   or    i.e.
- e.g., or e.g.

jwwojak replied:
changed to remove commas _and_ these should just be "See _name of thing_" and not "See also _name of thing_".

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -521,31 +523,31 @@ def touch_tip(
         """
         Touch the pipette tip to the sides of a well, with the intent of
-        removing left-over droplets
+        removing left-over droplets. See also, :ref:`touch-tip`.
```
closed compound. and the comma thing.
```suggestion
        removing leftover droplets. See also :ref:`touch-tip`.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -521,31 +523,31 @@ def touch_tip(
-        removing left-over droplets
+        removing left-over droplets. See also, :ref:`touch-tip`.
 
         :param location: If no location is passed, pipette will
```
would say "the pipette" here.

jwwojak replied:
😨 yikes. Fixing this. 

> :param location: If no location is passed, **the** pipette will . . .

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -521,31 +523,31 @@ def touch_tip(
-                         A positive offset moves the tip higher above the well,
-                         while a negative offset moves it lower into the well
-                         Default: -1.0 mm
+        :param v_offset: The offset in mm from the top of the well to touch tip.
```
Is it clearer to say something like `How far above or below the well to touch the tip, measured in mm.`?

jwwojak replied:
Yes. Better than just adding a period, which is all I did.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -521,31 +523,31 @@ def touch_tip(
         :type v_offset: float
         :param speed: The speed for touch tip motion, in mm/s.
-                      Default: 60.0 mm/s, Max: 80.0 mm/s, Min: 20.0 mm/s
+                      Default: 60.0 mm/s, Max: 80.0 mm/s, Min: 1.0 mm/s.
```
maybe better as bullets

jwwojak replied:
Yes, let's try it! Touchy syntax.

Text
New line
4 space indentation

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -521,31 +523,31 @@ def touch_tip(
+                      Default: 60.0 mm/s, Max: 80.0 mm/s, Min: 1.0 mm/s.
         :type speed: float
         :raises: ``UnexpectedTipRemovalError`` -- if no tip is attached to the pipette
         :raises RuntimeError: If no location is specified and location cache is
```
same as note on `blow_out()` above

jwwojak replied:
❓

jwwojak replied:
I'm not sure what to look for.

ecormany replied:
I think I was referring to making the parallel changes here in the explanation of the runtime error:

- [x] punctuate `eg`
- [x] double backticks around method name

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -602,25 +604,29 @@ def air_gap(
-                       to air-gap aspirate. (Default: 5 mm above current Well)
+
+        :param height: The height, in mm, to move above the current well
+                       before aspirating air. The default is 5 mm above current well.
```
terminology question here. we use `draw` as a more general verb to describe `asiprate()` above. it would make sense to mirror that here, instead of `pull`. 

i definitely want to avoid using `aspirate` as a verb for describing what `air_gap()` is doing. i think that introduces unnecessary confusion.

jwwojak replied:
This got hidden as "Outdated." Fixing now.

> Draw air into the pipette's tip at the current location. See :ref:`air-gap`.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -602,25 +604,29 @@ def air_gap(
 
         :raises: ``UnexpectedTipRemovalError`` -- if no tip is attached to the pipette
 
         :raises RuntimeError: If location cache is None.
```
same as other errors above

jwwojak replied:
Fixing.

> The amount of air to draw, measured in µL.

ecormany replied:
should actually say 
```
``air_gap`` 
```

jwwojak replied:
Updating text.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -654,7 +660,7 @@ def return_tip(self, home_after: Optional[bool] = None) -> InstrumentContext:
 
         This will not reset tip tracking, so the well flag will remain ``False``.
```
I think this means `has_tip`? But we never use the term `well flag` anywhere else, so we should track down what this means for sure and state it clearly.

jwwojak replied:
From @Laura-Danielle via Slack:
> The way that the protocol api determines whether a tip ‘exists’ in a tiprack is by this concept of a well_flag. If the protocol api sees that a well is marked as ‘False’ (i.e. empty) then it will not attempt to pick up from that location. Instead it will try to find a location that has a ‘True’ (i.e. filled) location in the tiprack.

ecormany replied:
it must be `has_tip`, that's the only bool in the Well class

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -681,8 +687,9 @@ def pick_up_tip(
 
-        If no location is passed, the Pipette will pick up the next available
+        If no location is passed, the pipette will pick up the next available
         tip in its :py:attr:`InstrumentContext.tip_racks` list.
```
prob add a tilde here for readability.

jwwojak replied:
Added ~ to shorten link to API reference:

> the pipette will pick up the next available tip in its :py:attr:`~.InstrumentContext.tip_racks` list.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
in `drop_tip()` there are several instances of the word Location surrounded by single backticks instead of double backticks. (lines 691-699)

jwwojak replied:
Yikes, italics.  _location_

Should be ``location``

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -700,8 +707,8 @@ def pick_up_tip(
           was not specified when creating the :py:class:`.InstrumentContext`.
```
the `tip_racks` property is mutable, so this isn't really about creation time so much as it's what the elements of the `tip_racks` list are right now.

jwwojak replied:
❓

ecormany replied:
maybe:
```
You can use this style of call to pick up a tip from a rack that is not a member of :py:obj:`.InstrumentContext.tip_racks`.
```

jwwojak replied:
Revised text.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -700,8 +707,8 @@ def pick_up_tip(
           was not specified when creating the :py:class:`.InstrumentContext`.
 
         * If you want to pick up the next available tip(s) in a specific
```
i'd love to kill the `(s)` if possible

jwwojak replied:
Oh, I hate those, the (s). Let's kill it with fire. Just use the plural even if something can be singular. Readers will understand. Save the (s) for the lawyers.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -700,8 +707,8 @@ def pick_up_tip(
+          tip rack, you may use the tip rack directly
+          (e.g. ``instr.pick_up_tip(tiprack)``).
 
         * If the position to move to in the well needs to be specified,
```
this is 🤯 wild. i wonder if anyone uses this, or if we even really want to support it. a matter for another day, probably.

nevertheless, `:type location:` needs to be edited to not display weirdly. `presses`, `increment`, and `prep_after` also need edits.

SyntaxColoring replied:
> this is 🤯 wild. i wonder if anyone uses this, or if we even really want to support it. a matter for another day, probably.

Picking up / dropping a tip with an offset, you mean?

In the Python Protocol API, I think it's helpful in environments like Jupyter Notebook where you want to be able to adjust things like that on-the-fly and experiment with different offsets. Outside the Python Protocol API, the system does this during Labware Position Check.

jwwojak replied:
I think these are ok now, though I'm lost among all the changes.

> nevertheless, :type location: needs to be edited to not display weirdly. presses, increment, and prep_after also need edits.

ecormany replied:
just remove `to pick up a tip from.` in line 727

jwwojak replied:
Removed, but I also don't see the exact quoted text. Very odd.  Lines 726-730:

> The number of times to lower and then raise the pipette when picking up a tip, to ensure a good seal (0 [zero] will result in the pipette hovering over the tip but not picking it up--generally not desirable, but could be used for dry-run).

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -681,8 +687,9 @@ def pick_up_tip(
 
-        If no location is passed, the Pipette will pick up the next available
+        If no location is passed, the pipette will pick up the next available
         tip in its :py:attr:`InstrumentContext.tip_racks` list.
```
also, all of this info about ``location`` should be _within_ the description of ``:param location:``.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -862,7 +869,7 @@ def drop_tip(
-        Drop the current tip.
+        Drop the current tip. See also, :ref:`pipette-drop-tip`.
 
         If no location is passed, the Pipette will drop the tip into its
```
```suggestion
        If no location is passed, the pipette will drop the tip into its
```
as above

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -862,7 +869,7 @@ def drop_tip(
+        Drop the current tip. See also, :ref:`pipette-drop-tip`.
 
         If no location is passed, the Pipette will drop the tip into its
         :py:attr:`trash_container`, which if not specified defaults to
```
the following text needs some Flex vs. OT-2 clarification

jwwojak replied:
@ecormany I wanted to tag you regarding this `drop_tip` part of the API reference.

`:param home_after:` has this _War and Peace_ length explanation. It goes into detail about problems with setting `home_after=false`. There's enough extra text this could be a standalone support article. Rather including this info in the API reference, I had this idea:

**Part 1**
For `:param home_after:` we say the following:

> Whether to home the pipette's plunger after dropping the tip. If not specified, defaults to ``True`` on an OT-2. Setting home_after=false is not recommended. See _link to new article TBD_.

(I don't know what a Flex does by default, yet.)

**Part 2**
Next, we take all the info about the problems setting `home_after=false` and put that in a new doc somewhere else. It doesn't have to be in our API docs. It could be a short support team article. I could create this and work with those guys to get it posted somewhere.

I'm recommending this because all the info makes `drop_tip` very long because of this (kinda obscure) argument that it accepts.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -877,19 +884,19 @@ def drop_tip(
 
             - If the only thing to specify is which well into which to drop
               a tip, `location` can be a :py:class:`.Well`. For instance,
```
more double backticks needed here, and a few lines up

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -877,19 +884,19 @@ def drop_tip(
+              ``instr.drop_tip(tiprack.wells()[0])``. This style of call can
               be used to make the robot drop a tip into arbitrary labware.
             - If the position to drop the tip from as well as the
               :py:class:`.Well` to drop the tip into needs to be specified,
```
this is such a weird and wordy phrasing. can definitely be tightened up.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -877,19 +884,19 @@ def drop_tip(
-              large height above the tiprack, `location`
+              large height above the tiprack, ``location``
               can be a :py:class:`.types.Location`; for instance, you can call
               `instr.drop_tip(tiprack.wells()[0].top())`.
```
is this unusually high? i would have assumed this was the default! we should find out and say what the default is.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -877,19 +884,19 @@ def drop_tip(
               :py:class:`.Well` to drop the tip into needs to be specified,
               for instance to tell the robot to drop a tip from an unusually
-              large height above the tiprack, `location`
+              large height above the tiprack, ``location``
```
```suggestion
              large height above the tip rack, ``location``
```
🧐

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -877,19 +884,19 @@ def drop_tip(
-            The location to drop the tip
+            The location to drop the tip.
         :type location:
             :py:class:`.types.Location` or :py:class:`.Well` or None
```
```suggestion
            :py:class:`.types.Location` or :py:class:`.Well` or ``None``
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -904,12 +911,12 @@ def drop_tip(
+                  pipette and tips.
                 * You understand the risks described below.
 
             The ejector shroud that pops the tip off the end of the pipette is
```
another shroud! let's send it to the grave too.
```suggestion
            The ejector that pops the tip off the end of the pipette is
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -980,7 +987,7 @@ def drop_tip(
     @requires_version(2, 0)
     def home(self) -> InstrumentContext:
-        """Home the robot.
+        """Home the robot. See also, :ref:`utility-homing`.
```
"Home the robot" sounds like it homes _everything_. Does the left pipette vertical axis home if you call this on the right pipette? OK to follow up later if we don't have a quick answer.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1342,7 +1349,7 @@ def move_to(
     @requires_version(2, 0)
     def mount(self) -> str:
-        """Return the name of the mount this pipette is attached to"""
+        """Return the name of the mount this pipette is attached to."""
```
I presume the only possibilities are `left` and `right`? Might be worth listing if so.

jwwojak replied:
Yes, checked on that. Will add.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1458,7 +1465,7 @@ def trash_container(self, trash: labware.Labware) -> None:
     def name(self) -> str:
         """
-        The name string for the pipette (e.g. 'p300_single')
+        The name string for the pipette (e.g. 'p300_single).
```
```suggestion
        The name string for the pipette (e.g., ``p300_single``).
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1479,10 +1486,10 @@ def min_volume(self) -> float:
     @requires_version(2, 0)
     def max_volume(self) -> float:
```
we should also document `min_volume()` above, unless it's not supported for some reason.

jwwojak replied:
Checked in Slack (python-protocol-api). Seth says the missing text is probably an oversight. I will add a minimum description.

ecormany replied:
let's also add to the min definition:
```
This value may change based on the :ref:`volume mode <FILL IN ANCHOR>` that the pipette is currently configured for.
```

jwwojak replied:
Added text and link to volume mode.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1524,7 +1531,7 @@ def hw_pipette(self) -> PipetteDict:
     @requires_version(2, 0)
     def channels(self) -> int:
-        """The number of channels on the pipette."""
+        """The number of channels on the pipette. See also, :ref:`new-pipette`."""
```
add that it's 1, 8, or 96?

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1524,7 +1531,7 @@ def hw_pipette(self) -> PipetteDict:
+        """The number of channels on the pipette. See also, :ref:`new-pipette`."""
         return self._core.get_channels()
 
     @property  # type: ignore
```
i'd never heard of this, but it's just a getter, so it's probably fine to leave as is. except `tiprack` in line 1540.

jwwojak replied:
Changed "tiprack" to "tip rack."

jwwojak replied:
Checking throughout file for instances of "tiprack." Will only change text, not code (e.g. `tiprack`, or `if parent_labware.is_tiprack`)

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1584,27 +1591,21 @@ def __str__(self) -> str:
-
-        The Flex 1-Channel 50 µL and Flex 8-Channel 50 µL pipettes must operate in a low-volume mode
-        to accurately dispense 1 µL of liquid. Low-volume mode can only be set by calling this function.
+        """Configure a pipette to handle a specific volume of liquid, measured in µL. Depending on the volume, the pipette will enter a unique pipetting mode.
```
not sure i like "unique" here. there's two modes, and that makes it sound like there's a lot more. stet or `The pipette enters a volume mode depending on the volume provided.`

jwwojak replied:
Revised using the recommended text.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1584,27 +1591,21 @@ def __str__(self) -> str:
 
-        For more information on available modes for certain pipettes, see :ref:`pipette-volume-modes`.
+        The Flex 1-Channel 50 µL and Flex 8-Channel 50 µL pipettes must operate in a low-volume mode to accurately dispense 1 µL of liquid. Low-volume mode can
+        only be set by calling ``configure_for_volume``. See also, :ref:`pipette-volume-modes`.
```
i like the demonstrative pronoun better. indicates "you are already here". i could see reading this and thinking "oh i have to look up another function"

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1584,27 +1591,21 @@ def __str__(self) -> str:
         .. note ::
 
-            Changing a pipette's mode will reset its :ref:`flow rates <new-plunger-flow-rates>`.
+            Changing a pipette's mode will reset its :ref:`flow rate <new-plunger-flow-rates>`.
```
stet. all the flow rates (aspirate, dispense, blowout). yes, these happen to be identical for the Flex pipettes in question, but that could change.

jwwojak replied:
leaving text as is.

ecormany replied:
oops, i meant leave the original: `rates`

jwwojak replied:
stetted 😉

ecormany reviewed (commented):
This is big, so I left lots of comments. Mostly small stuff on the edited parts. A few methods and parameters were passed over entirely and need a first round of edits.

jwwojak committed:
Apply suggestions from code review

Making batch commits before other changes. Will do a batch commit, pull locally, more revisions, and push again.

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak committed:
Remove commas in introductory signals

For text like "See, . . . " and "See also, . . . " let's get rid of the commas.

Also, most, if not all, of these should be just "See" not "See also" because we're referring to stuff that directly supports the text.

jwwojak committed:
Style code formatting, text, and code sample

Single quotes made code italics. Replace with ``code`` to get monospace.

Remove * as a multiplication operator. Replace with "multiplied by".

Fix code sample in dispense.

Run Black check. Looks ok.

SyntaxColoring commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -125,20 +124,21 @@ def starting_tip(self, location: Union[labware.Well, None]) -> None:
     @requires_version(2, 0)
     def reset_tipracks(self) -> None:
-        """Reload all tips in each tip rack and reset starting tip"""
+        """Reloads all tips in each tip rack and resets the starting tip."""
```
The first line of Python docstrings is conventionally imperative ("reload"/"reset" instead of "reloads"/"resets"), unfortunately.

See [PEP 257](https://peps.python.org/pep-0257/#:~:text=The%20docstring%20is%20a%20phrase%20ending%20in%20a%20period.%20It%20prescribes%20the%20function%20or%20method%E2%80%99s%20effect%20as%20a%20command%20(%E2%80%9CDo%20this%E2%80%9D%2C%20%E2%80%9CReturn%20that%E2%80%9D)%2C%20not%20as%20a%20description%3B%20e.g.%20don%E2%80%99t%20write%20%E2%80%9CReturns%20the%20pathname%20%E2%80%A6%E2%80%9D.).

ecormany replied:
Welp, standards gonna standard. We can comply 🫡

ecormany replied:
🧐 linguistics nerd alert 🤓 
i think they're bare vPs, not imperative clauses, but whatever. i've [ticketed](https://opentrons.atlassian.net/browse/RTC-359) this for a followup, since i spotted some other TP clauses in the docstrings of other classes (e.g., `HeaterShakerContext`):

jwwojak replied:
🖖 fixing this.

jwwojak committed:
Revisions from reviewers part 1

Incorporating reviewer comments.
- aspirate
- dispense
- reset_tipracks
- touch_tip
- remove (s)
- small adjustments

Not finished yet. More to come.

jwwojak committed:
Revisions from comments

- Replace (text only) "tiprack" with "tip rack"
- Fix single quotes around code elements. Replace with double single quotes (`not this`, but ``this instead``). Rendered text in italics.
- drop_tip revisions
- mount revisions, add "left" and "right"
- min_volume revision, add definition
- configure_for_volume changes

jwwojak committed:
Fogot to save recent and run Black

jwwojak committed:
Fix code snippet in touch_tip

Another instance of italics instead of monospace.

jwwojak committed:
Everything but the intro

jwwojak committed:
Revised intro

jwwojak commented:
I've seen formatting issues in the labware class too, but that's out of scope for this PR, which covers the InstrumentContext.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -55,18 +55,17 @@
 
 
 class InstrumentContext(publisher.CommandPublisher):
-    """A context for a specific pipette or instrument.
```
We have to leave the one-line summary. This is a Python docstring format standard.

jwwojak replied:
Restored orignal intro sentence.

> A context for a specific pipette or instrument.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -55,18 +55,17 @@
+       The InstrumentContext class in the Opentrons Python API (PAPI) provides the objects,
+       attributes, and methods that allow you to use pipettes and other instruments in your
+       protocols. Similar to other class containers, InstrumentContext is temporary.
+       Class objects only exist when instantiated in your Python protocol.
```
A couple things:
- "PAPI" is internal shorthand only.
- "temporary" is maybe too strong — you can't switch pipettes during a protocol, so they persist for its duration

Maybe replace this ¶ and the next one:
```suggestion
    """
       The InstrumentContext class provides the objects,
       attributes, and methods that allow you to use pipettes in your protocols. 
       
       Methods generally fall into one of two categories. 
       
         - They can change the state of the InstrumentContext object, such as how fast 
         it moves liquid or where it disposes of used tips). 
         - They can command the instrument to perform an action, like picking up tips, moving
         to certain locations, and aspirating or dispensing liquid. 
```
check my formatting, i composed this directly here in github.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -154,26 +153,28 @@ def aspirate(
-        this pipette.
+        Draw liquid into a pipette tip. Includes parameters that control
+        aspiration volume, well location, tip position in a well,and pipette flow rate.
+        See :ref:`new-aspirate`.
```
Summarizing the immediately following parameter list is redundant. And let's keep the first line special.
```suggestion
        Draw liquid into a pipette tip.
        
        See :ref:`new-aspirate` for more details and examples.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -154,26 +153,28 @@ def aspirate(
+                              well bottom is specified by the :py:obj:`well_bottom_clearance.aspirate` object.
+
+                            - If the location is a ``location`` (e.g., the result of :py:meth:`.Well.top`
+                              or :py:meth:`.Well.bottom`), the robot will aspirate from that specified position.
```
the classes `Well` and `Location` should be capitalized. it wouldn't matter overly much, except here lowercase `location` is the name of the _parameter_ in question. (naming things is hard.)
```suggestion
                            - If the location is a ``Well``, the robot will aspirate at or above
                              the bottom center of the well. The distance (in mm) from the
                              well bottom is specified by :py:obj:`well_bottom_clearance.aspirate`.

                            - If the location is a ``Location`` (e.g., the result of :py:meth:`.Well.top`
                              or :py:meth:`.Well.bottom`), the robot will aspirate from that specified position.
```
🐍 🤓 `well_bottom_clearance.aspirate` isn't an object, it's a parameter of an object.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -183,7 +184,7 @@ def aspirate(
             to guess whether the argument is a volume or location - it is
```
This note could be more succinct, but it's fine.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -256,35 +257,41 @@ def dispense(
-        into the specified location.
+        Dispense liquid from a pipette tip. Includes parameters that control
+        dispense volume, well location, tip position in a well, and pipette
+        flow rate. See :ref:`new-dispense`.
```
like aspirate above
```suggestion
        Dispense liquid from a pipette tip.

        See :ref:`new-dispense` for more details and examples.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -256,35 +257,41 @@ def dispense(
-        :param volume: The volume of liquid to dispense, in microliters. If 0
-                       or unspecified, defaults to :py:attr:`current_volume`.
-        :type volume: int or float
+                            - If the location is a ``well``, the pipette will dispense at or above
```
Same as with aspirate above

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -256,35 +257,41 @@ def dispense(
+                              object.
+
+                            - If the location is a ``location`` (e.g., the result of :py:meth:`.Well.top` or
+                              :py:meth:`.Well.bottom`), the robot  will dispense into the specified position.
```
```suggestion
                              :py:meth:`.Well.bottom`), the robot will dispense into the specified position.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -377,25 +384,25 @@ def mix(
-                     and when dispensing, it will be
-                     `rate` * :py:attr:`flow_rate.dispense <flow_rate>`.
+        Mixes a volume of liquid by repeatedly aspirating and dispensing it in a single location.
+        See :ref:`mix`.
```
```suggestion
        Mix a volume of liquid by repeatedly aspirating and dispensing it in a single location.
        
        See :ref:`mix`.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -377,25 +384,25 @@ def mix(
+        :param volume: The volume to mix, measured in µL. If 0 or unspecified,
+                       defaults to the maximum volume for the pipette and its
+                       attached tip.
+        :param location: A well or a position relative to well
```
It would probably be good to use the `Well or Location` terminology from other methods.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -445,25 +452,24 @@ def blow_out(
     ) -> InstrumentContext:
         """
-        Blow liquid out of the tip.
+        Blows an extra amount of air through a pipette's tip to clear it.
```
```suggestion
        Blow an extra amount of air through a pipette's tip to clear it.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -445,25 +452,24 @@ def blow_out(
 
         :raises RuntimeError: If no location is specified and location cache is
-                              None. This should happen if `blow_out` is called
+                              None. This should happen if ``blow_out`` is called
```
```suggestion
                              ``None``. This should happen if ``blow_out`` is called
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -484,7 +490,7 @@ def blow_out(
             if target.well.parent.is_tiprack:
                 _log.warning(
-                    "Blow_out being performed on a tiprack. "
+                    "Blow_out being performed on a tip rack. "
```
This isn't a docstring and could have an effect on tests. Revert.
```suggestion
                    "Blow_out being performed on a tiprack. "
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -521,31 +527,34 @@ def touch_tip(
+
+                        - Default: 60.0 mm/s
+                        - Max: 80.0 mm/s
+                        - Min: 1.0 mm/s
```
I like spelling these out.
```suggestion
                        - Maximum: 80.0 mm/s
                        - Minimum: 1.0 mm/s
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -521,31 +527,34 @@ def touch_tip(
         :raises: ``UnexpectedTipRemovalError`` -- if no tip is attached to the pipette
         :raises RuntimeError: If no location is specified and location cache is
-                              None. This should happen if `touch_tip` is called
+                              None. This should happen if ``touch_tip`` is called
```
```suggestion
                              ``None``. This should happen if ``touch_tip`` is called
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -579,7 +588,7 @@ def touch_tip(
         if parent_labware.is_tiprack:
             _log.warning(
-                "Touch_tip being performed on a tiprack. Please re-check your code"
+                "Touch_tip being performed on a tip rack. Please re-check your code"
```
Not a docstring, could affect tests. Revert.
```suggestion
                "Touch_tip being performed on a tiprack. Please re-check your code"
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -602,25 +611,28 @@ def air_gap(
     ) -> InstrumentContext:
         """
-        Pull air into the pipette current tip at the current location
+        Draw air into the pipette's tip at the current location. See :ref:`air-gap`.
```
Param description make clear this happens at the same _well_ but not at the same exact location (i.e., the pipette moves to get out of the well so it doesn't air-gap in a location full of liquid).
```suggestion
        Draw air into the pipette's tip at the current well.

        See :ref:`air-gap`.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -602,25 +611,28 @@ def air_gap(
 
-        :param volume: The amount in uL to aspirate air into the tube.
-                       (Default will use all remaining volume in tip)
+        :param volume: The ``air_gap``, measured in µL.
```
Too circular.
```suggestion
        :param volume: The amount of air, measured in µL.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -862,75 +871,46 @@ def drop_tip(
     ) -> InstrumentContext:
         """
-        Drop the current tip.
+        Drop the current tip. See :ref:`pipette-drop-tip`.
```
```suggestion
        Drop the current tip.

        See :ref:`pipette-drop-tip`.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -862,75 +871,46 @@ def drop_tip(
-        in order to prevent tips from piling up in a single location in the trash.
+        If no location is passed (e.g. ``pipette.drop_tip()``), the pipette will drop the
+        attached tip into its default :py:attr:`trash_container`. For the Flex and OT-2 the
+        default slots for their in-deck trash containers are A3 and 12, respectively.
```
i wouldn't characterize the OT-2 as having an "in-deck" trash container.
```suggestion
        default slots for their trash containers are A3 and 12, respectively.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -980,7 +960,7 @@ def drop_tip(
     @requires_version(2, 0)
     def home(self) -> InstrumentContext:
-        """Home the robot.
+        """Home the robot. See :ref:`utility-homing`.
```
```suggestion
        """Home the robot.

        See :ref:`utility-homing`.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1342,7 +1322,10 @@ def move_to(
     def mount(self) -> str:
-        """Return the name of the mount this pipette is attached to"""
+        """
+        Return the name of the mount the pipette is attached to. The names are
+        ``left`` and ``right``.
```
```suggestion
        Return the name of the mount the pipette is attached to.

        The possible names are ``left`` and ``right``.
```

ecormany replied:
note on which the 96-channel reports?

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1388,7 +1371,7 @@ def speed(self) -> "PlungerSpeeds":
     @requires_version(2, 0)
     def flow_rate(self) -> "FlowRates":
-        """The speeds (in uL/s) configured for the pipette.
+        """The speeds, in µL/s, configured for the pipette. See :ref:`new-plunger-flow-rates`.
```
```suggestion
        """The speeds, in µL/s, configured for the pipette.

        See :ref:`new-plunger-flow-rates`.
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1403,7 +1386,7 @@ def flow_rate(self) -> "FlowRates":
 
         For instance, to change the flow rate for aspiration on an instrument
-        you would do
+        you would do:
```
i think we can delete this whole example since we added a cross-reference

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1516,21 +1504,21 @@ def _has_tip(self) -> bool:
     @requires_version(2, 0)
     def channels(self) -> int:
-        """The number of channels on the pipette."""
+        """The number of channels on the pipette. See :ref:`new-pipette`."""
```
```suggestion
        """The number of channels on the pipette.

        See :ref:`new-pipette`."""
```

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1584,27 +1572,26 @@ def __str__(self) -> str:
         .. note ::
 
-            Changing a pipette's mode will reset its :ref:`flow rates <new-plunger-flow-rates>`.
+            Changing a pipette's mode will reset its :ref:`flow rate <new-plunger-flow-rates>`.
```
```suggestion
            Changing a pipette's mode will reset its :ref:`flow rates <new-plunger-flow-rates>`.
```

ecormany reviewed (approved):
The substance is good; all that remains is pedantry! I've left many comments to make a record of why certain changes are necessary. I'll batch up the suggestions into a commit and then also do a final pass through to clean up line lengths. Then this will be ready to merge.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1458,7 +1441,7 @@ def trash_container(self, trash: labware.Labware) -> None:
     def name(self) -> str:
         """
-        The name string for the pipette (e.g. 'p300_single')
+        The name string for the pipette (e.g., ``p300_single``).
```
```suggestion
        The name string for the pipette (e.g., ``"p300_single"``).
```
string value

ecormany committed:
batch of changes from final review

jwwojak committed:
Restore intro sentence to line 58

and run Black.

All done! ✨ 🍰 ✨
1 file reformatted.
